draft-ietf-ppm-dap.md

```diff
@@ -1267,106 +1334,173 @@ This message consists of:
+Leader MUST abort the aggregation job.
+
+Otherwise, the Leader proceeds as follows with each report:
+
+1. If the inbound prep step has type "continue", then the Leader computes
+
+   ~~~
+   (state, outbound) = VDAF.ping_pong_req(agg_param, state, inbound)
+   ~~~
+
+   where `inbound` is the message payload. If `outbound != None`, then the
+   Leader stores `state` and `outbound` and proceeds to
+   {{aggregation-leader-continuation}}. If `outbound == None`, then the
+   preparation process is complete: either `state == Rejected()`, in which case
+   the Leader rejects the report and removes it from the candidate set; or
+   `state == Finished(out_share)`, in which case preparation is complete and the
+   Leader stores the output share for use in the collection sub-protocol
+   {{collect-flow}}.
+
+1. Else if the type is "rejected", then the Leader rejects the report and
+   removes it from the candidate set. Unless the error is `report_too_early`,
+   the Leader MUST NOT include the report in a subsequent aggregation job.
+   Otherwise, the Leader MAY include the report in a subsequent aggregation
+   job.
+
+1. Else the type is invalid, in which case the Leader MUST abort the
+   aggregation job.
+
+(Note: Since VDAF preparation completes in a constant number of rounds, it will
+never be the case that some reports are completed and others are not.)
 
 #### Helper Initialization {#aggregation-helper-init}
 
-Each Helper begins their portion of the aggregate initialization when they
-receive an AggregationJobInitReq message from the Leader. For each ReportShare
+The Helper begins an aggregation job when it receives an
+`AggregationJobInitReq` message from the Leader. For each `PrepareStep`
 conveyed by this message, the Helper attempts to initialize VDAF preparation
 (see {{Section 5.1 of !VDAF}}) just as the Leader does. If successful, it
-includes its prepare message in its response that the Leader will use to
-continue the process.
+includes the result in its response that the Leader will use to continue
+preparing the report.
 
 To begin this process, the Helper first checks if it recognizes the task ID. If
-not, then it MUST abort with error `unrecognizedTask`. Then the Helper checks
-that the report IDs in `AggregationJobInitReq.report_shares` are all distinct.
-If two ReportShare values have the same report ID, then the helper MUST abort
-with error `invalidMessage`. If this check succeeds, the helper then attempts to
-recover each input share in `AggregationJobInitReq.report_shares` as follows:
+not, then it MUST abort with error `unrecognizedTask`.
+
+Next, the Helper checks that the report IDs in
+`AggregationJobInitReq.prepare_steps` are all distinct. If two prep step
+messages have the same report ID, then the Helper MUST abort with error
+`invalidMessage`.
+
+Next, the Helper checks that each prep step message has type "initialize". If
+not, it MUST abort with error `invalidMessage`.
```

Comment:
If it's illegal for the `prepare_steps` in `AggregationJobInitReq` to ever be anything but `PrepareStep::initialize`, why bother with the enum at all? We could instead have something like `struct PrepareStepInit { ReportShare report_share; opaque payload<0..2^32-1>; }` and remove one possible source of bugs in implementations.

Reply:
I agree--in fact, splitting things into separate messages for initialization/continuation was the initial approach taken in this PR--but we decided to merge in order to be able to use the same types for initialization & continuation and to minimize the size of the PR. (https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/393#discussion_r1117814328)

I'd prefer something like you suggest (that is, step-specific message types), but I would like consensus before making a change. @cjpatton?

Reply:
Note that there are still more edge cases like this that we can't eliminate as easily.

Reply:
I introduced `PrepareInit` and `PrepareContinue` as messages that the Leader sends to the helper for initialization & continuation steps.

IMO, `PrepareStep`'s name is now iffy -- perhaps it should be something like `PrepareResponse`? -- but I think we can rename in a later PR, if we decide we want to.